### DISCUSSION TITLE
Any document about embedding immudb?

### DISCUSSION THREAD

**Discussion by iredmail** [codenotary/immudb#1041] [Q&A]
Dear developers,

Where can i find more info about embedding immudb in a Go project?
Besides some issues mention using the "embedded" package, no other info on website and github repo.

**Comment by jeroiraz** [comment]
Hi @iredmail, thanks for reaching out.
Some brief documentation can be found at https://docs.immudb.io/master/develop/embedding.html

I'm adding a section as embedded key-value store, I'll place the link into this issue

**Comment by iredmail** [comment]
@jeroiraz Thanks for helping. :)

btw:

- is it possible to just use the store and not the sql package?
- any plan to move the whole (or sub-directories of) `embedded/` directory as a separate git repo and go package?

**Comment by jeroiraz** [comment]
yes, it's possible to use the embedded key-value store

```
st, err := store.Open("data", store.DefaultOptions())
// handle error
defer st.Close()

tx, err := st.Commit([]*store.KV{
{Key: []byte("hello"), Value: []byte("immutable-world!")},
}, true)
// handle error

fmt.Printf("key '%s' successfully set at tx %d", "hello", tx.ID)

value, txID, nupdates, err := st.Get([]byte("hello"))
// handle error
```

No plans to move to another git repo the embedded packages, but they can be used without  importing server-related code, not even sql if just importing the embedded/store package

**Comment by iredmail** [comment]
Would you mind updating document to mention what features are offered by the store? Compare to the immudb daemon/service, what features are losing (like SQL, of course), and what features are still there, is it still immutable, etc.

Currently the `embedded/store/` doesn't have any README or api document, no comments for functions/methods, have to go through all source lines to figure it out.

I guess it requires much time to document it, and i just ask for too much. :(

**Comment by jeroiraz** [comment]
Actually the embedded store has all the core functionality, including SQL (optional when the sql engine is used as described in https://docs.immudb.io/master/develop/embedding.html.

The server provides user management, multi-database, built-in master-follower replication, state signing, and exposing the endpoints to be consumed by SDKs.

It will depend on the application you're developing, if you need/want an embedded store or use it as an independent component.

**Comment by iredmail** [comment]
Dear @jeroiraz 

Thanks. I will try to embed the `store` package in my own project.

btw, It would be great either the source code can has more comment lines, or a detailed document for store package. :)

**Comment by iredmail** [comment]
A stupid question: In `IndexOptions`, what does the `Thld` mean in option `FlushThld` and `CompactionThld`? 
All other attributes in `Options` and `IndexOptions` use full English word in names, except this `Thld`. :(
Maybe we should rename it to make it clearer?

**Comment by jeroiraz** [comment]
`Thld` stands for `threshold`

`FlushThld`, after specified number of insertions the changes are flushed to disk
`CompactionThld`, index compaction can take place after specified number of snapshots has been flushed into disk

**Comment by iredmail** [comment]
How to do auditing with embedded kv store? Like `immuclient audit-mode` command.
